### ISSUE TITLE
interceptRequest fails with "The arguments passed to a command are either invalid or malformed"

### ISSUE THREAD

**Issue by adrian-brown** [webdriverio/webdriverio#5611]
[//]: # NOTE: This repository only maintains packages that are listed in the [Readme](https://github.com/webdriverio/webdriverio/blob/master/README.md#packages). Please make sure that your issue is directly caused by one of these packages and if not file an issue in the correct 3rd party package repository.

**Environment (please complete the following information):**
 - **WebdriverIO version:** [6.1.25]
 - **Mode:** [WDIO Testrunner]
 - **If WDIO Testrunner, running sync/async:** [sync]
 - **Node.js version:** [10.17.0]
 - **NPM version:** [6.13.4]
 - **Browser name and version:** [Chrome 79]
 - **Platform name and version:** [Windows 10]
 - **Additional wdio packages used (if applicable):** [@wdio/sauce-service]

**Config of WebdriverIO**
N/A

**Describe the bug**
Using `interceptRequest()` fails on SauceLabs with error:
The arguments passed to a command are either invalid or malformed.

**To Reproduce**
Use this

```
     browser.interceptRequest({
        url: 'https://saucelabs.com',
        redirect: 'https://google.com'
      })
```
[Include code or an example repository that can easily be set up]

**Expected behavior**
This request should intercept the URL request for foo and send to bar.  It should not error out.

**Log**
If applicable, add logs to help explain your problem.
Create a [gist](https://gist.github.com) which is a paste of your _full_ Selenium/WebdriverIO logs, and link them here. Do _not_ paste your full Appium logs here, as it will make this issue very long and hard to read! If you are reporting a bug, _always_ include Selenium/WebdriverIO logs!

**Additional context**

> Please use the correct markdown

**Comment by christian-bromann** [comment]
@adrian-brown thanks for reporting. The issue has been tracked internally at Sauce Labs and will be tackled by the team soon.

**Comment by nescalante** [comment]
Is there a way to track the progress of this issue in the Sauce Labs backlog?

**Comment by christian-bromann** [comment]
@nescalante there hasn't been any updates yet, please subscribe to this issue for more information on resolutions to this bug.

**Comment by ayroblu** [comment]
@christian-bromann Does it work in specific cases? It seems like the examples on the website don't work, and so this feature doesn't exist, and a waste of time for others to try out?

**Comment by antonborisoff** [comment]
@adrian-brown , @nescalante , @ayroblu , it seems the documentation is invalid. I contacted the guys in SauceLabs and they said it is available in the following form:
```
await browser.execute('sauce:intercept', {
  url: 'https://www.saucedemo.com/v1/index.html',
  redirect: 'https://saucelabs.com/',
});
```
You need to set extendedDebugging to true. Here is the link to docs (ignore interceptRequest command in the examples, it doesn't work, you should execute the script): https://docs.saucelabs.com/insights/debug/

**Comment by christian-bromann** [comment]
@antonborisoff thanks for letting us know. So does:

```ts
await browser.interceptRequest({
  url: 'https://www.saucedemo.com/v1/index.html',
  redirect: 'https://saucelabs.com/',
});
```

not work anymore?

**Comment by wswebcreation** [comment]
@christian-bromann 

No, it doesn't. I'm checking internally what the issue is. This works 
![image](https://user-images.githubusercontent.com/11979740/179201658-10b3b4c5-4bac-48d1-b3ae-834abc81e25c.png)

but this doesn't
![image](https://user-images.githubusercontent.com/11979740/179201690-eae514b2-446a-4bf4-8841-cbb54001f6e9.png)

**Comment by christian-bromann** [comment]
Maybe:
```ts
await browser.interceptRequest('https://www.saucedemo.com/v1/index.html', 'https://saucelabs.com/');
```
?

**Comment by wswebcreation** [comment]
nope, that fails in WDIO because it expects a rule object

**Comment by christian-bromann** [comment]
@wswebcreation can you check by tweaking the protocol definition in `/packages/wdio-protocols/apis/sauce.json` I think the problem is that the protocol was not defined correctly in WebdriverIO.

**Comment by wswebcreation** [comment]
@christian-bromann 

When I change this in the `saucelabs.json` file 

```json
      "parameters": [{
        "name": "rule",
        "type": "object",
        "description": "A rule describing the request interception.",
        "required": true
      }]
```

to (don't mind the description

```json
      "parameters": [{
        "name": "url",
        "type": "string",
        "description": "A rule describing the request interception.",
        "required": true
      },{
        "name": "redirect",
        "type": "string",
        "description": "A rule describing the request interception.",
        "required": true
      }]
    }
  },
```

it works with this command

```js
    await browser.interceptRequest(
      'https://www.saucedemo.com/v1/index.html',
      'https://saucelabs.com'
    );
```

**Comment by christian-bromann** [comment]
That's it!

**Comment by wswebcreation** [comment]
Shall I create a PR for this? It's a pretty easy fix, but also needs a website release

By the way, I like this

```js
    await browser.interceptRequest({
      url: 'https://www.saucedemo.com/v1/index.html',
      redirect: 'https://saucelabs.com',
    });
```
better then this
```js
    await browser.interceptRequest(
      'https://www.saucedemo.com/v1/index.html',
      'https://saucelabs.com'
    );
```
Not sure if I create a PR for it where this needs to be changed.

**Comment by christian-bromann** [comment]
> Not sure if I create a PR for it where this needs to be changed.

It would be awesome if you could send a PR. A change to the protocol file should be sufficient.

> By the way, I like this

Unfortunately protocol commands are currently implemented so that every body parameter is a function parameter. I suggest to implement Sauce Labs support for the current [mock object](https://webdriver.io/docs/api/browser/mock) so that local testing works the same way as remote testing.

**Comment by wswebcreation** [comment]
ok, will first create a PR for it

**Comment by wswebcreation** [comment]
Hmm, @christian-bromann 

It might not be as easy, see https://docs.saucelabs.com/insights/debug/#intercept-network-requests

There are multiple options:

- redirect
- error
- response

How should I handle that in the protocol?

**Comment by wswebcreation** [comment]
ok, let me try that locally. Not sure If I'm able to create a PR for it today, but will let you know

**Comment by ayroblu** [comment]
I saw that webdriverio 7 supports mock, I'm not sure what value saucelabs intercept has over webdriver's mock

**Comment by christian-bromann** [comment]
@ayroblu the mocking capabilities webdriverio currently has are only for local testing and don't work with Sauce Labs. This command allows similar features for running tests in the cloud.

**Comment by ayroblu** [comment]
Thanks, I see, I guess if we end up using this we'll need to write a wrapper around them

**Comment by wswebcreation** [comment]
Tnx @christian-bromann 
Will prepare a PR for it later today

**Comment by AlbertORoura** [comment]
Hello @wswebcreation how is going on that PR?

Any body knows if there is any other option to intercept request?

Would be appreciated if anybody can bring some light on that.

Thank you very much.

**Comment by AlbertORoura** [comment]
@wswebcreation @christian-bromann any news?

**Comment by wswebcreation** [comment]
Sorry, no not really.

This is the endpoint

```json
"/session/:sessionId/sauce/ondemand/intercept": {
    "POST": [{
      "command": "interceptRequest",
      "description": "Allows modifying any request made by the browser. You can blacklist, modify, or redirect these as required for your tests.",
      "ref": "https://wiki.saucelabs.com/display/DOCS/Custom+Sauce+Labs+WebDriver+Extensions+for+Network+and+Log+Commands#CustomSauceLabsWebDriverExtensionsforNetworkandLogCommands-InterceptNetworkRequests",
      "examples": [
        [
          "// redirect a request",
          "await browser.interceptRequest({",
          "  'https://saucelabs.com',",
          "  'https://google.com'",
          "})"
        ],
        [
          "// Modify requests to REST API (Mock REST API response)",
          "await browser.interceptRequest({",
          "  'http://sampleapp.appspot.com/api/todos',",
          "  null,",
          "  {",
          "    headers: {",
          "      'x-custom-headers': 'foobar'",
          "    },",
          "    body: [{",
          "      title: 'My custom todo',",
          "      order: 1,",
          "      completed: false,",
          "      url: 'http://todo-backend-express.herokuapp.com/15727'",
          "    }]",
          "  }",
          "})"
        ],
        [
          "// Blacklist requests to 3rd party vendors",
          "await browser.interceptRequest({",
          "  'https://api.segment.io/v1/p',",
          "   null,",
          "  'Failed'",
          "})"
        ]
      ],
      "parameters": [{
        "name": "url",
        "type": "string",
        "description": "An outgoing request URL. Wildcards are supported, for example https://saucelabs.com/rest/v1/* would redirect any v1 rest API requests.",
        "required": true
      },{
         "name": "redirect",
        "type": "string",
        "description": "An absolute URL to where the original request is redirected.",
        "required": false

      }]
    }]
  },
```

- having 3 commands in here seems to be impossible
- having 3 times the same endpoint is also not a thing
- adding optional arguments also gives issues

seems like I’m stuck here, any ideas would be appreciated.

Temporary workaround would be not to use the API provided by the service but the "normal" options that Sauce Labs offers with [extendedDebugging](https://docs.saucelabs.com/insights/debug/#using-the-javascript-executor-to-simulate-network-conditions)

**Comment by christian-bromann** [comment]
Closing due to inactivity.
